### PR TITLE
Prevent embedding of non-abstract interface members.

### LINKED ISSUE #35852: VB DIM: An attempt to embed an interface with non-abstract members should cause an error
```
        <Fact>
        Public Sub NoPia_02()
            Dim attributesRef = GetCSharpCompiation(NoPiaAttributes).EmitToImageReference()

            Dim csSource =
"
using System;
using System.Runtime.InteropServices;
using System.Runtime.CompilerServices;

[assembly: PrimaryInteropAssemblyAttribute(1,1)]
[assembly: Guid(""f9c2d51d-4f44-45f0-9eda-c9d599b58257"")]

[ComImport()]
[Guid(""f9c2d51d-4f44-45f0-9eda-c9d599b58279"")]
public interface ITest33
{
    void M1(){}
}
" + NoPiaAttributes

            Dim csCompiation = GetCSharpCompiation(csSource, {attributesRef}).EmitToImageReference(embedInteropTypes:=True)

            Dim source1 =
<compilation>
    <file name="c.vb"><![CDATA[
class UsePia 
    Sub Main(x as ITest33)
        x.M1()
    End Sub
End Class
]]></file>
</compilation>

            Dim comp1 = CreateCompilation(source1, targetFramework:=TargetFramework.NetStandardLatest, references:={attributesRef, csCompiation})
            ' Expect an error similar to - CS8711: Type 'ITest33' cannot be embedded because it has a non-abstract member. Consider setting the 'Embed Interop Types' property to false.
            comp1.AssertTheseDiagnostics()
        End Sub
```

### LINKED ISSUE #35911: An attempt to embed an interface with a re-abstraction of a method from base should produce an error
```
        [Fact]
        public void NoPia_10()
        {
            var attributes = CreateCompilation(NoPiaAttributes, options: TestOptions.ReleaseDll, targetFramework: TargetFramework.NetStandardLatest);
            var attributesRef = attributes.EmitToImageReference();

            string pia = @"
using System;
using System.Runtime.InteropServices;
using System.Runtime.CompilerServices;

[assembly: PrimaryInteropAssemblyAttribute(1,1)]
[assembly: Guid(""f9c2d51d-4f44-45f0-9eda-c9d599b58257"")]

[ComImport()]
[Guid(""f9c2d51d-4f44-45f0-9eda-c9d599b58279"")]
public interface ITest33
{
    void M1();
}

[ComImport()]
[Guid(""f9c2d51d-4f44-45f0-9eda-c9d599b58280"")]
public interface ITest44 : ITest33
{
    abstract void ITest33.M1();
}
";

            var piaCompilation = CreateCompilation(pia, options: TestOptions.ReleaseDll, references: new[] { attributesRef }, targetFramework: TargetFramework.NetStandardLatest);

            CompileAndVerify(piaCompilation, verify: VerifyOnMonoOrCoreClr);

            string consumer = @"
class UsePia
{
    public static void Main(ITest44 x)
    {
        x.M1();
    }
} 
";

            foreach (var reference in new[] { piaCompilation.ToMetadataReference(embedInteropTypes: true), piaCompilation.EmitToImageReference(embedInteropTypes: true) })
            {
                var compilation1 = CreateCompilation(consumer, options: TestOptions.ReleaseDll, references: new[] { reference, attributesRef }, targetFramework: TargetFramework.NetStandardLatest);

                compilation1.VerifyEmitDiagnostics(
                    // (4,29): error CS8711: Type 'ITest44' cannot be embedded because it has a non-abstract member. Consider setting the 'Embed Interop Types' property to false.
                    //     public static void Main(ITest44 x)
                    Diagnostic(ErrorCode.ERR_DefaultInterfaceImplementationInNoPIAType, "ITest44").WithArguments("ITest44").WithLocation(4, 29)
                    );
            }
        }
```

### PR DESCRIPTION
**Pull request by AlekseyTs**
Closes #35852. Closes #35911.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Emitter/NoPia/EmbeddedTypesManager.cs [hunk 1]
```diff
@@ -233,10 +233,18 @@ internal override void ReportIndirectReferencesToLinkedAssemblies(AssemblySymbol
...
                     {
-                        if (member.Kind != SymbolKind.NamedType && !member.IsAbstract)
+                        if (member.Kind != SymbolKind.NamedType)
                         {
-                            error = ErrorCode.ERR_DefaultInterfaceImplementationInNoPIAType;
-                            break;
+                            if (!member.IsAbstract)
+                            {
+                                error = ErrorCode.ERR_DefaultInterfaceImplementationInNoPIAType;
+                                break;
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Emitter/NoPia/EmbeddedTypesManager.cs]: This does not feel like the right behavior for a constant declared in an interface.  Similarly for a static method.  I'm not sure we intend to forbid them in an embedded interface, but even if we do this seems like the wrong diagnostic.

**Comment by gafter** [comment]
>                     if (error != ErrorCode.Unknown)

Shouldn't we also check that it inherits nothing inappropriate from its base interfaces?

---
Refers to: src/Compilers/CSharp/Portable/Emitter/NoPia/EmbeddedTypesManager.cs:251 in f46c209. [](commit_id = f46c209828a1e31f082586bffc1989562dc279dc, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Errors/ErrorCode.cs [hunk 2]
```diff
@@ -1725,10 +1725,12 @@ internal enum ErrorCode
         ERR_NotNullConstraintMustBeFirst = 8713,
         WRN_NullabilityMismatchInTypeParameterNotNullConstraint = 8714,
 
-        ERR_DuplicateNullSuppression = 8715
+        ERR_DuplicateNullSuppression = 8715,
 
         #endregion diagnostics introduced for C# 8.0
 
+        ERR_ReAbstractionInNoPIAType = 8750,
+
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Errors/ErrorCode.cs]: I think this belongs above the #endregion marker above, as it is being introduced for C# 8.0.

File: src/Compilers/CSharp/Portable/Emitter/NoPia/EmbeddedTypesManager.cs (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Emitter/NoPia/EmbeddedTypesManager.cs]: > This does not feel like the right behavior for a constant declared in an interface.  Similarly for a static method.  

NoPia feature is not supposed to embed constants or static methods. BTW, constants are never referenced explicitly in metadata, therefore they do not come through this code path.

> I'm not sure we intend to forbid them in an embedded interface, but even if we do this seems like the wrong diagnostic.

Could you please elaborate what is wrong with it?

---
In reply to: [315919836](https://github.com/dotnet/roslyn/pull/38140#discussion_r315919836) [](ancestors = 315919836)

**Comment by AlekseyTs** [comment]
>                     if (error != ErrorCode.Unknown)

> Shouldn't we also check that it inherits nothing inappropriate from its base interfaces?

If a base interface is supposed to be embedded, it will be checked on its own. If it is not supposed to be embedded, we do not care what is inside it for the purpose of NoPia.

---
In reply to: [523210234](https://github.com/dotnet/roslyn/pull/38140#issuecomment-523210234) [](ancestors = 523210234)

---
Refers to: src/Compilers/CSharp/Portable/Emitter/NoPia/EmbeddedTypesManager.cs:251 in f46c209. [](commit_id = f46c209828a1e31f082586bffc1989562dc279dc, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Errors/ErrorCode.cs (hunk 2 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Errors/ErrorCode.cs]: > I think this belongs above the #endregion marker above, as it is being introduced for C# 8.0.

Moved

---
In reply to: [315921033](https://github.com/dotnet/roslyn/pull/38140#discussion_r315921033) [](ancestors = 315921033)

**Pull request merged**